### PR TITLE
Add Formless component template file

### LINKED ISSUE #49: Add a component template file/directory?
Similar to what I did with [`ComponentTemplate.purs`](https://github.com/JordanMartinez/learn-halogen/blob/latestRelease/src/05-ComponentTemplate-A.purs), would you be up for me adding something like that here? Or should I perhaps store that in my learning repo?

Due to how much stuff goes on in creating a form, I think this should be a directory with these files:
- data and validation: defines a single type in the form, its instances, and its validation code
- component - commented: a Formless component template file that explains everything (useful for new learners)
- component - uncommented: a Formless component template file that explains nothing and is used to create the boilerplate to make the thing work (useful for developers)

### PR DESCRIPTION
**Pull request by JordanMartinez**
## What does this pull request do?

Fixes #49

Also changes `halogen-storybook`'s branch from `halogen5` to `master` as the code wouldn't compile otherwise.

## Where should the reviewer start?

Read the 'other notes' section, skim the code and then answer my questions below

## How should this be manually tested?

AFAIK, we should only verify that this code compiles in the CI. 

## Other Notes:

There's a number of things that aren't correct in this PR with regard to the data type, `Name`.
- the output field of the row type isn't `Name` as it should be but `String`
- there might be a few type classes that `Name` unnecessarily implements (can't remember)
- ~there isn't an optic for `FieldError`, so the usage of `_Error` will likely stop compilation~
- there's no example of the `Initial` type class and how one would could implement it for `Name`

The halogen component part is basically the uncommented version of the template file. Since it's also lacking in a few other ways (more modular pieces that show how to use some of its API?), it would need to be completed before we could document it better. I'm not sure what should be included here and what should be left out because the `examples` folder already exists.

Questions I have:
- What else is `st.form` used for?
- I think the commented version should include 'additional state' as an example of its usage whereas the uncommented one should not. Your thoughts?
- I think this file should be separately licensed to ensure that one who uses it (since that's what it's for, right?) don't have to worry about satisfying such license constraints. Your thoughts?

Lastly, there's documenting each part in a clear manner.

### CODE REVIEW

**Comment by JordanMartinez** [comment]
@thomashoneyman Mind taking a look at this sometime soon so I can finish it up?

File: template/FormlessTemplate.purs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+import Data.Newtype (class Newtype)
+import Effect.Aff (Aff)
+import Formless as F
+import Halogen.HTML as HH
+import Halogen.HTML.Events as HE
+import Halogen.HTML.Properties as HP
+import Template.DataAndValidation (NAME_FIELD, _name, minLength)
+import Type.Row (type (+))
+
+type FormFieldsRow f =
```
Comment (thomashoneyman) [inline template/FormlessTemplate.purs:16]: I quite like this!

File: template/FormlessTemplate.purs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+      , HH.button
+        [ if st.submitting || st.validity /= F.Valid
+            then HP.disabled true
+            else HE.onClick \_ -> Just F.submit
+        ]
+        [ HH.text "Submit" ]
+      ]
+
+  handleEvent :: F.Event Form AddedState
+              -> F.HalogenM Form AddedState Action ChildSlots Message MonadType Unit
```
Comment (thomashoneyman) [inline template/FormlessTemplate.purs]: Mind dropping these down for indentation?

```purs
  handleEvent 
    :: F.Event Form AddedState
    -> F.HalogenM Form AddedState Action ChildSlots Message MonadType Unit
```

File: template/FormlessTemplate.purs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+  handleEvent = case _ of
+    F.Submitted formContent -> do
+      -- This is how to get the output values of the form.
+      let formFields = F.unwrapOutputFields formContent
+
+      -- We won't do this here, but this is how most will handle a form
+      -- submission: raise it as an event to their parent.
+      --      H.raise formFiels
```
Comment (thomashoneyman) [inline template/FormlessTemplate.purs]: formFields*

File: template/FormlessTemplate.purs [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+  | Receive Input
+
+data Query a
+  = Command a
+  | Reply (Unit -> a)
+
+type Message = FormFields
+type ChildSlots =
+  ()
+type MonadType = Aff
```
Comment (thomashoneyman) [inline template/FormlessTemplate.purs]: I think I'd rather this remain polymorphic, `m`, as that's the most common way the component is used in practice. Then just `component :: forall m. MonadAff m => ... m`

File: template/FormlessTemplate.purs [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+component :: F.Component Form Query ChildSlots Input FormFields Aff
+component = F.component mkInput $ F.defaultSpec
+  { render = render
+  , handleAction = handleAction
+  , handleEvent = handleEvent
+  , handleQuery = handleQuery
+  , initialize = Just Initialize
+  , finalize = Just Finalize
+  , receive = Just <<< Receive
+  }
```
Comment (thomashoneyman) [inline template/FormlessTemplate.purs:57]: I think each of these (`mkInput` included) could use a one-line description of what it's for, just as a reminder. If you would like, I can add those in for you, or feel free to do so. 

Something simple like:

```purs
-- Decide what, if anything, to do when Formless events occur. If you would like to raise 
-- events as messages, then use `F.raiseResult` as your `handleEvent` function.
handleEvent :: ...
```

File: template/FormlessTemplate.purs [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+      , HH.p_ [ HH.text $ "Number of Submit attempts: " <> show st.submitAttempts ]
+      , HH.p_ [ HH.text $ "Additional state was: " <> show st.additionalState ]
+
+      , HH.div_
+        [ HH.text "Label" ]
+      , HH.input
+        [ HP.type_ InputText
+        , HP.placeholder "Michael"
+        , HP.value (F.getInput _name st.form) -- access one value in form
+        , HE.onValueInput (Just <<< F.setValidate _name)
```
Comment (thomashoneyman) [inline template/FormlessTemplate.purs:113]: Perhaps a comment here, too:

```suggestion
        , HE.onValueInput (Just <<< F.setValidate _name) -- set the value of _name, then validate it
```

**Comment by thomashoneyman** [comment]
Thanks for adding this @JordanMartinez -- I'm happy to merge with a couple of small tweaks I mentioned above. This is quite nice. I'll debug the CI failure, but it's not to do with your PR and so I am happy to merge even if I haven't quite fixed that yet.

**Comment by JordanMartinez** [comment]
I hope to get to this sometime tonight.

**Comment by JordanMartinez** [comment]
@thomashoneyman I've addressed all your comments except for documenting more parts. I think that might be better handled by you since you're more familiar with this library than I am.

File: template/FormlessTemplate.purs [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    -- everything else below comes from our `AddedState` rows:
+    , additionalState: Just 5
+    }
+
+  render
+    :: F.PublicState Form AddedState
+    -> F.ComponentHTML Form Action ChildSlots m
+  render st =
+    HH.div_
+      [ HH.p_ [ HH.text $ "Validity: " <> show st.validity ]
```
Comment (JordanMartinez) [inline template/FormlessTemplate.purs:80]: Indicates whether the form's values are valid (i.e. validation has passed for all fields)

File: template/FormlessTemplate.purs [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+      , HH.p_ [ HH.text $ "Dirty: " <> show st.dirty ]
```
Comment (JordanMartinez) [inline template/FormlessTemplate.purs:84]: Indicates whether the form is currently being modified/validated.

File: template/FormlessTemplate.purs [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+      , HH.p_ [ HH.text $ "Being Submitted: " <> show st.submitting ]
```
Comment (JordanMartinez) [inline template/FormlessTemplate.purs:89]: Indicates whether the 'submit' button has been clicked and the form's content is still being validated one last time before submission is accepted.

File: template/FormlessTemplate.purs [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+      , HH.p_ [ HH.text $ "Number of Errors: " <> show st.errors ]
```
Comment (JordanMartinez) [inline template/FormlessTemplate.purs:93]: Indicates the number of errors due to validation failing that need to be fixed before submission is accepted

File: template/FormlessTemplate.purs [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+      , HH.p_ [ HH.text $ "Number of Submit attempts: " <> show st.submitAttempts ]
```
Comment (JordanMartinez) [inline template/FormlessTemplate.purs:96]: Indicates the number of times user has attempted to submit the form

File: template/FormlessTemplate.purs [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+      , HH.p_ [ HH.text $ "Additional state was: " <> show st.additionalState ]
```
Comment (JordanMartinez) [inline template/FormlessTemplate.purs:101]: This is how we refer to the additional labels we added to the form's state in `AddedState`.

File: template/FormlessTemplate.purs [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+      ]
+
+  -- Decide what, if anything, to do when Formless events occur.
+  -- For example, if you would like to raise events as messages,
+  -- then use `F.raiseResult` as your `handleEvent` function.
+  handleEvent
+    :: F.Event Form AddedState
+    -> F.HalogenM Form AddedState Action ChildSlots Message m Unit
+  handleEvent = case _ of
+    F.Submitted formContent -> do
```
Comment (JordanMartinez) [inline template/FormlessTemplate.purs:131]: Indicates that the form has been successfully submitted.

File: template/FormlessTemplate.purs [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+
+      -- We won't do this here, but this is how most will handle a form
+      -- submission: raise it as an event to their parent.
+      --      H.raise formFields
+
+      -- Alternatively, one could do something custom with the output values.
+
+      -- This line exists so the code compiles.
+      pure unit
+    F.Changed formState -> do
```
Comment (JordanMartinez) [inline template/FormlessTemplate.purs:147]: Indicates that the form's content has been changed. This event is triggered anytime a field is changed, whether it passes validation or not.

**Review by JordanMartinez — COMMENTED** [review]
I'm not sure what you had in mind for adding comments to each part, but this is what I would put. Feel free to correct/revise, etc. and I'll update it to what we agree upon here and push the changes following that.

File: template/FormlessTemplate.purs (hunk 8 above)
Comment (thomashoneyman) [reply template/FormlessTemplate.purs:84]: Indicates whether any field in the form has been changed from its initial state

File: template/FormlessTemplate.purs (hunk 12 above)
Comment (thomashoneyman) [reply template/FormlessTemplate.purs:101]: We can also refer to any additional labels we used to extend the form's state; in this case, that means any field from our `AddedState` type.

**Comment by thomashoneyman** [comment]
I've added a thumbs-up to your suggestions which I think are good as-is, and I've added my own version for ones that I'd tweak. On the whole you're almost spot-on with what I'd write. Thanks!

**Comment by thomashoneyman** [comment]
CI has been fixed thanks to @linearray in #54, so once your updates here are applied I'll merge this in. Thanks again!

**Comment by JordanMartinez** [comment]
Done! Just waiting on CI now.

**Comment by thomashoneyman** [comment]
@JordanMartinez mind rebasing or merging the current `halogen5` branch into yours to get the CI passing? Thanks!

**Comment by thomashoneyman** [comment]
Actually -- don't worry about it, I'll just merge and let things run there. Thanks again!

**Pull request merged**